[docs] Improved latency description in APISocketOptions (#1607)
Haivision/srt: docs/APISocketOptions.md

| [`SRTO_CONNTIMEO`](#SRTO_CONNTIMEO)                    | 1.1.2 | pre     | `int32_t` | ms      | 3000          | 0..      | W   | GSD+  |
| [`SRTO_DRIFTTRACER`](#SRTO_DRIFTTRACER)                | 1.4.2 | post    | `bool`    |         | true          |          | RW  | GSD   |
| [`SRTO_ENFORCEDENCRYPTION`](#SRTO_ENFORCEDENCRYPTION)  | 1.3.2 | pre     | `bool`    |         | true          |          | W   | GSD   |
| [`SRTO_EVENT`](#SRTO_EVENT)                            |       |         | `int32_t` | flags   |               |          | R   | S     |
| [`SRTO_FC`](#SRTO_FC)                                  |       | pre     | `int32_t` | pkts    | 25600         | 32..     | RW  | GSD   |
| [`SRTO_GROUPCONNECT`](#SRTO_GROUPCONNECT)              | 1.5.0 | pre     | `int32_t` |         | 0             | 0...1    | W   | S     |
| [`SRTO_GROUPSTABTIMEO`](#SRTO_GROUPSTABTIMEO)          | 1.5.0 | pre     | `int32_t` | ms      | 80            | 10-...   | W   | GSD+  |
| [`SRTO_GROUPTYPE`](#SRTO_GROUPTYPE)                    | 1.5.0 | pre     | `int32_t` | enum    |               |          | R   | S     |

> | [`SRTO_INPUTBW`](#SRTO_INPUTBW)                        | 1.0.5 | post    | `int64_t` | B/s     | 0             | 0..      | RW  | GSD   |

Reviewer: B/s = bits/s?

Author: Bytes per second. Bits are usually denoted by lower case "b".

| [`SRTO_IPV6ONLY`](#SRTO_IPV6ONLY)                      | 1.4.0 | pre     | `int32_t` |         | (system)      | -1..1    | RW  | GSD   |
| [`SRTO_ISN`](#SRTO_ISN)                                | 1.3.0 |         | `int32_t` |         |               |          | R   | S     |
| [`SRTO_KMPREANNOUNCE`](#SRTO_KMPREANNOUNCE)            | 1.3.2 | pre     | `int32_t` | pkts    | 0x1000        | 0.. *    | RW  | GSD   |
| [`SRTO_KMREFRESHRATE`](#SRTO_KMREFRESHRATE)            | 1.3.2 | pre     | `int32_t` | pkts    | 0x1000000     | 0..      | RW  | GSD   |
| [`SRTO_KMSTATE`](#SRTO_KMSTATE)                        | 1.0.2 |         | `int32_t` | enum    |               |          | R   | S     |
| [`SRTO_LATENCY`](#SRTO_LATENCY)                        | 1.0.2 | pre     | `int32_t` | ms      | 120 *         | 0..      | RW  | GSD   |
| [`SRTO_LINGER`](#SRTO_LINGER)                          |       | pre     | `linger`  | s       | on, 180       | 0..      | RW  | GSD   |
| [`SRTO_LOSSMAXTTL`](#SRTO_LOSSMAXTTL)                  | 1.2.0 | pre     | `int32_t` | packets | 0             | 0..      | RW  | GSD+  |

> | [`SRTO_MAXBW`](#SRTO_MAXBW)                            | 1.0.5 | pre     | `int64_t` | B/s     | -1            | -1..     | RW  | GSD   |

Reviewer: B/s = bits/s?

Author: Maybe then Bytes/s for better readability... 🤔

| --------------------- | ----- | ------- | ---------- | ------ | -------- | ------ | --- | ------ |
| `SRTO_LATENCY`        | 1.0.2 | pre     | `int32_t`  | ms     | 120 *    | 0..    | RW  | GSD    |

- This option sets both [`SRTO_RCVLATENCY`](#SRTO_RCVLATENCY) and [`SRTO_PEERLATENCY`](#SRTO_PEERLATENCY)
to the same value specified.

- Prior to SRT version 1.3.0 `SRTO_LATENCY` was the only option to set the latency.
However it is effectively equivalent to setting `SRTO_PEERLATENCY` on the sender 
(see [`SRTO_SENDER`](#SRTO_SENDER)), and `SRTO_RCVLATENCY` on the receiver.
SRT version 1.3.0 and higher support bidirectional transmission, so that each side can 

> be sender and receiver at the same time, and `SRTO_SENDER` became redundant.

Reviewer 2: That's not the main reason to have two different options PEERLATENCY/RCVLATENCY. The reason is that since 1.3.0 and bidirectional support, the latency setting is **per direction** - that exactly thing should be mentioned.

Author: Changed to
> However it is effectively equivalent to setting `SRTO_PEERLATENCY` in the sending direction
(see [`SRTO_SENDER`](#SRTO_SENDER)), and `SRTO_RCVLATENCY` in the receiving direction.

> - Prior to SRT version 1.3.0 `SRTO_LATENCY` was the only option to set the latency.
> However it is effectively equivalent to setting `SRTO_PEERLATENCY`, when the side is 
> sender (see [`SRTO_SENDER`](#SRTO_SENDER)), and `SRTO_RCVLATENCY` when the side is receiver.
> SRT version 1.3.0 and higher support bidirectional transmission, so that each sides can be sender and receiver
> at the same time, and the `SRTO_SENDER` became redundant.

Reviewer: ```suggestion
- Prior to SRT version 1.3.0 `SRTO_LATENCY` was the only option to set the latency.
However it is effectively equivalent to setting `SRTO_PEERLATENCY` on the sender 
(see [`SRTO_SENDER`](#SRTO_SENDER)), and `SRTO_RCVLATENCY` on the receiver.
SRT version 1.3.0 and higher support bidirectional transmission, so that each side can 
be sender and receiver at the same time, and `SRTO_SENDER` became redundant.
```

| -------------------- | ----- | ------- | ---------- | ------- | -------- | ------ | --- | ------ |
| `SRTO_MAXBW`         | 1.0.5 | pre     | `int64_t`  | B/s     | -1       | -1..   | RW  | GSD    |

- Maximum send bandwidth:
  - `-1`: infinite (the limit in Live Mode is 1 Gbps);
  - ` 0`: relative to input rate (see [`SRTO_INPUTBW`](#SRTO_INPUTBW));

>   - `>0`: absolute limit in B/s.

Reviewer: B/s = bits/s?

Author: Bytes per second.

Reviewer 2: I think it could be mentioned in the beginning, before the whole option table, that all options that refer to transmission speed or bandwidth limit, are usually expressed in bytes per second, denoted as B/s.

Author: Added to the legend:
> - `B/s` - bytes per second.

| -------------------- | ----- | ------- | ---------- | ------- | -------- | ------ | --- | ------ |
| `SRTO_PEERLATENCY`   | 1.3.0 | pre     | `int32_t`  | ms      | 0        | 0..    | RW  | GSD    |

- The latency value (as described in [`SRTO_RCVLATENCY`](#SRTO_RCVLATENCY)) provided by the sender 
side as a minimum value for the receiver.

> - Reading the value of the option before on a non-connected socket reports the configured value.

Reviewer: ```suggestion
- Reading the value of the option on an unconnected socket reports the configured value.
```

| OptName           | Since | Binding | Type       |  Units  |   Default  | Range  | Dir | Entity |
| ----------------- | ----- | ------- | ---------- | ------- | ---------- | ------ | --- | ------ |
| `SRTO_RCVLATENCY` | 1.3.0 | pre     | `int32_t`  | ms      | *          | 0..    | RW  | GSD    |

> - The latency value in the receiving direction of the current SRT connection.

Reviewer 2: Well, "of the current SRT connection" is not the happiest statement here, as it might be understood that you should set this option upon connection, which is actually the opposite of how you should use it.

Maybe `The latency value in the receiving direction to be used for this socket's connection` ?

Author: Changed to
> The latency value in the receiving direction of the socket.

This value is only significant when [`SRTO_TSBPDMODE`](#SRTO_TSBPDMODE) is enabled.

- **Default value**: 120 ms (depicted as 0) in Live mode, 0 in File mode (see [`SRTO_TRANSTYPE`](#SRTO_TRANSTYPE)).

> - The latency value defines the **minimum** receiver buffering delay before delivering an SRT data packet

Reviewer 2: The use of `minimum` is kinda unfortunate here. This is minimum from the perspective of negotiation of the effective value, not from the perspective of applying this value to the received data, while it sounds here as the latter.

You can also check this in the code, but as far as I remember how it works is that the transmission start time latch happens at the moment of receiving the first data packet. The time when it was received for that very packet is simply treated as the expected reception time, and the time from the timestamp is then used to define the delta used to calculate the value of the expected reception time for the next packets. Then expected reception time + latency + dfift = play time. I might describe it also as:

First packet:
```
dTp = [packet's timestamp];
Texp = [now];  <-- here we state that expected arrival time is when it was arrived
Tstart = Texp - dTp;  <-- here we define Tstart the very first time
Tplay = Texp + [latency] + [drift];
```
Next packets:
```
dTp = [packet's timestamp];
dTs = [collected time from over-32bit time overflows];
Texp = Tstart + dTs + dTp;  <-- here we can use Tstart we defined previously
Tplay = Texp + [latency] + [drift];
```

Author: > The use of minimum is kinda unfortunate here. This is minimum from the perspective of negotiation of the effective value, not from the perspective of applying this value to the received data, while it sounds here as the latter.

The word "minimum" is used here for the reading delay (TSBPD), as the actual delay can be bigger than the one set on the socket.

> The time when it was received for that very packet is simply treated as the expected reception time, and the time from the timestamp is then used to define the delta used to calculate the value of the expected reception time for the next packets. Then expected reception time + latency + dfift = play time. I might describe it also as:

From the sender's and the whole transmission point of view, "the time when it was received" gives additional RTT/2 to the end-to-end latency, as described.

> - The latency value defines the **minimum** receiver buffering delay before delivering an SRT data packet
> from receiving SRT socket to a receiving application. The provided value is used on the connection establishment (handshake exchange) stage
> to fixate the end to end latency of the transmission. The effective end-to-end latency `L` will be fixed
> aa the network transmission time of the final handshake packet (~1/2 RTT) plus the **negotiated** latency value `Ln`.
> Data packets will stay in the receiver buffer for at least `L` microseconds since the timestamp of the
> packet independent of the actual network transmission times (RTT variations) of these packets.

Reviewer: ```suggestion
- The latency value defines the **minimum** receiver buffering delay before delivering an SRT data packet
from a receiving SRT socket to a receiving application. The provided value is used in the connection establishment (handshake exchange) stage
to fix the end-to-end latency of the transmission. The effective end-to-end latency `L` will be fixed
as the network transmission time of the final handshake packet (~1/2 RTT) plus the **negotiated** latency value `Ln`.
Data packets will stay in the receiver buffer for at least `L` microseconds since the timestamp of the
packet, independent of the actual network transmission times (RTT variations) of these packets.
```

- The actual value of the receiver buffering delay `Ln` (the negotiated latency) used on a connection

> is determined by the negotiation on the connection establishment (handshake exchange) phase as the maximum of the

Reviewer: ```suggestion
is determined by the negotiation in the connection establishment (handshake exchange) phase as the maximum of the
```

`SRTO_RCVLATENCY` value and the value of [`SRTO_PEERLATENCY`](#SRTO_PEERLATENCY) set by the peer.

- Reading the `SRTO_RCVLATENCY` value on a socket after the connection is established provides the actual (negotiated)
latency value `Ln`.

- The receiver's buffer must be large enough to store the `L` segment of the stream,
i.e. `L × Bitrate` bytes. Refer to [`SRTO_RCVBUF`](#SRTO_RCVBUF).

- The sender's buffer must be large enough to store a packet up until it is either delivered (and acknowledged)

> or dropped by the sender due to too late to be delivered.

Reviewer: ```suggestion
or dropped by the sender due to it becoming too late to be delivered.
```
